modules/dhcp_common/subnet_service.rb

```diff
@@ -79,7 +79,8 @@ def add_lease(subnet_address, record)
 
     def add_host(subnet_address, record)
       m.synchronize do
-        reservations_by_ip[subnet_address, record.ip] = record
+        reservations_by_ip[subnet_address, record.ip] ||= []
+        reservations_by_ip[subnet_address, record.ip] << record
```

Comment:
Alternatively, you could change MemoryStore constructor to accept a default value for new keys.